### ISSUE TITLE
QAOA

### ISSUE THREAD

**Issue by DiegoGM91** [Quantum-TII/qibo#213]
As discussed in the meeting this morning, we could have the QAOA built-in in qibo, which may attract potential users. Here is my code (if you find it acceptable as a first version, I shall open a pull request):

``import numpy as np``
``from qibo import models, hamiltonians``
``from qibo import optimizers # this is not currently used``

``class QAOA(object):``
   `` """ This class implements the QAOA algorithm."""``
        
    def __init__(self, nqubits, hamiltonian, mixing=None, initial_state=None):
        """
        Initializes an instance of the QAOA.
        
        Args:
            nqubits (int): number of qubits.
            hamiltonian (qibo.hamiltonians.Hamiltonian): problem Hamiltonian whose ground state is sought.
            mixing (qibo.hamiltonians.Hamiltonian): mixing Hamiltonian. If None, it is taken as -sum(sigma_x).
            initial state: initial state of the algorithm. If None, it is taken as the uniform
                           superposition over computational-basis states.
        """
        # problem hamiltonian
        self.hamiltonian = hamiltonian
        # mixing hamiltonian (default = -sum(sigma_x))
        if mixing == None:
            self.mixing = hamiltonians.X(nqubits)
        else:
            self.mixing = mixing
        # initial state
        if initial_state == None:
            self.initial_state = np.ones(2**nqubits) / np.sqrt(2**nqubits)
        else:
            self.initial_state = initial_state
            
        
    def loss(self, p):
        """
        Args: 
            p (numpy.array or list): parameters of the QAOA.
        Returns:
            (numpy.float64) value of the loss function.
        """
        # We define the evolution operators for the mixing and problem hamiltonians 
        evolve = []
        for i in range(len(p)//2):
            evolve.append(models.StateEvolution(self.hamiltonian, p[2*i]))
            evolve.append(models.StateEvolution(self.mixing, p[2*i+1]))
        # We evolve the state
        state = self.initial_state
        for i,evolution_operator in enumerate(evolve):
            state = evolution_operator(final_time=p[i], initial_state=state)

        return np.float64(self.hamiltonian.expectation(state))
    
    
    def train(self, initial_p):
        """
        Args:
            initial_p (numpy.array or list): initial guess for the parameters of the QAOA.
        Returns:
            (float, numpy.array) with  (minimum found for the problem hamiltonian, optimal angles)
        """
        from scipy.optimize import minimize

        def positivity_constraint(p):
            """
            Args:
                p (numpy.array or list): parameters of the QAOA.
            Returns: 
                0 iff all parameters are non-negative
            """
            return np.sum(np.sign(p))-len(p)
        
        cons = {'type':'eq', 'fun': positivity_constraint}
        print('Optimizing QAOA...')
        result = minimize(self.loss, initial_p, method='SLSQP', constraints=cons)
        
        return result.fun, result.x 

At the moment, time evolution in qibo must have positive time increments, otherwise the code throws an exception. This is why this snippet of code implements a restricted or constrained version of the QAOA. My view is that there is nothing wrong with negative time increments in time evolution, as this simply means evolving backwards in time. As Stavros pointed out, this is irrelevant for the time evolution model itself, because it suffices to multiply the Hamiltonian by -1. However, it does matter if one is to develop applications like the QAOA, where the Hamiltonian is fixed. I don't know what your views on this are.

The SLSQP optimizer however does not seem to work, as it returns the same initial parameters right away. I've seen on the internet that other people run into this very same issue with this minimizer, but none of the proposed solutions worked here. So now I'm trying to understand how the mystic library works. Still, I decided to share the code so that you can tell me what you think (on the other hand, if we allow for negative time increments, there is no need to use constrained optimization).
